sentinelhub/data_utils.py

```diff
@@ -116,37 +126,46 @@ def statistical_to_dataframe(result_data: List[JsonDict], exclude_stats: Optiona
     return pandas.concat(dfs)
 
 
-def _get_failed_intervals(
-    identifier: str, response_data: List[JsonDict]
-) -> Optional[Dict[str, Union[str, List[Tuple[str, str]]]]]:
-    """Collect failed intervals of a partially failed request
+def _get_failed_intervals(result_data: JsonDict) -> Union[str, List[Tuple[str, str]]]:
+    """Collect failed intervals of a single geometry from the (Batch) Statistical result
 
-    :param identifier: The identifier of the geometry.
-    :param response_data: An input representation of Statistical API response.
-    :return: The identifier of a geometry that has a response status of PARTIAL and the failed intervals.
+    :param result_data: An input representation of the (Batch) Statistical API result of a geometry.
+    :return: The failed intervals for a geometry.
     """
-    failed_intervals = []
-    for interval in response_data:
-        if "error" in interval:
-            failed_intervals.append((interval["interval"]["from"], interval["interval"]["to"]))
-
-    return {"identifier": identifier, "failed_intervals": failed_intervals} if failed_intervals else None
-
-
-def get_failed_statistical_requests(result_data: List[JsonDict]) -> List[Dict[str, Union[str, List[Tuple[str, str]]]]]:
+    if _is_batch_stat(result_data):
+        return (
+            "full time range"
+            if "error" in result_data or not result_data["response"]
+            else [
+                (interval["interval"]["from"], interval["interval"]["to"])
+                for interval in result_data["response"]["data"]
+                if "error" in interval
+            ]
+        )
+
+    return (
+        "full time range"
+        if "error" in result_data
+        else [
+            (interval["interval"]["from"], interval["interval"]["to"])
+            for interval in result_data["data"]
+            if "error" in interval
+        ]
+    )
+
+
+def get_failed_statistical_requests(result_data: List[JsonDict]) -> List[JsonDict]:
     """Collect failed requests of (Batch) Statistical Results
 
     :param result_data: An input representation of (Batch) Statistical API result.
     :return: Failed requests of (Batch) Statistical Results.
     """
-    failed_requests = []
-    for result in result_data:
-        identifier, response = result["identifier"], result["response"]
-        if not response:
-            failed_requests.append({"identifier": identifier})
-        else:
-            failed_intervals = _get_failed_intervals(identifier, response["data"])
-            if failed_intervals:
-                failed_requests.append(failed_intervals)
-
-    return failed_requests
+    if _is_batch_stat(result_data[0]):
+        return [
+            {"identifier": result["identifier"], "failed_intervals": _get_failed_intervals(result)}
+            for result in result_data
+            if _get_failed_intervals(result)
+        ]
```

Comment:
This doesn't seem quite right. `_get_failed_intervals` returns a string or a tuple, so why are you doing `if _get_failed_intervals(result)`?

Reply:
It returns a string, a list of tuples, or an empty list if there's no error in any intervals.

Reply:
ah, right. Make it more explicit with `if _get_failed_intervals(result) != []` or something. We have had many issues due to not being specific with `if x` instead of `if x is None` and similar.

Reply:
I am also not a fan of very "busy" list comprehensions, and prefer doing them in steps. How does this sound to you:
```python
failed_intervals = ((result["identifier"], _get_failed_intervals(result)) for result in result_data)
    return [
        {"identifier": identifier, "failed_intervals": intervals}
        for identifier, intervals in failed_intervals
        if intervals != []
    ]
```

Reply:
I'm actually confused when to/ not to use list comprehensions atm. I feel it is a bit conflicting with avoid `if else`, `continue`, `break`, etc. Is there a general guild line for this?

Reply:
If you have a more difficult control flow (e.g. multiple `if`s or `for` loops) in a list comprehension, which spans multiple lines, it should be refactored. That does not mean not to use list comprehension, it just means you need more steps.

When in doubt, i suggest you write out both options and choose the one that is more readable. Readability > anything else (unless performance is a priority).

In this case we have
```python
failed_intervals = []
for result in result_data:
    intervals_for_result = _get_failed_intervals(result)
    if intervals_for_result != []:
        failed_intervals.append({"identifier": result["identifier"], "failed_intervals": intervals_for_result})

return failed_intervals
```
versus

```python
        return [
            {"identifier": result["identifier"], "failed_intervals": _get_failed_intervals(result)}
            for result in result_data
            if _get_failed_intervals(result)
        ]
```

You can see that the first one has more lines of code, requires more variable names. And it tells you *how* to build the result.
The second one tells you more about *what* is being returned. In general that is a bit favored for readability. But IMO large list comprehensions suffer from the same problem, where we just effectively move all the control flow inside a list comprehension. In that case it is probably better to do it in multiple steps.

```python
    failed_intervals = ((result["identifier"], _get_failed_intervals(result)) for result in result_data)
    return [
        {"identifier": ident, "failed_intervals": intervals} for ident, intervals in failed_intervals if intervals != []
    ]
```

Here I tried to separate two steps. The first one calculates failed intervals for all the results. The second one filters and repackages them.

But in the end a lot of this is subjective.